### DISCUSSION TITLE
Yes you can create classes on the stack!

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#98] [General]
# Yes you can create classes on the stack!

In this article we will create a class, aka a reference type on the stack, and therefore don't use any managed memory!

https://steven-giesel.com/blogPost/74cc8abc-b6e3-4cb0-ba98-9c92f9a5a4d2/yes-you-can-create-classes-on-the-stack

**Comment by toupswork** [comment]
Wow that's so cool. I never thought of that. Still, I'd stick with using structs since they don't have the extra overhead of a class, such as the header with it's sync block and pointers to the virtual function table etc. Plus the wasted space for alignment of that header

**Reply by linkdotnet** [reply]
I also don’t have any use case where this could be useful at all.
As you said - use a struct.

**Comment by james-allison** [comment]
Not sure this is right - I think:
var instance = Unsafe.AsRef<MyClass>(mem);

should be:
var instance = Unsafe.AsRef<MyClass>(&mem);

In answer to when it might be useful - I've some elderly API's I'm using that take byte[] rather than Span<byte> - so was exploring being able to make short lived byte[]'s without heap allocation. (Sure it'll end badly...)

**Reply by linkdotnet** [reply]
Thanks for noticing that @james-allison - yes it should be `&mem` as `mem` is basically the pointer to the pointer (aka the pointer to the `TypeHandle` value).

> so was exploring being able to make short lived byte[]'s without heap allocation.

At least that is relatively easy in modern C# versions with `Span<byte> stackAllocedArray = stackalloc byte[256};`.

**Reply by james-allison** [reply]
No problem at all - many thanks for writing this up!

I ended up making an extension method for ReadOnlySpan<byte> to create an 'unsafe' byte[] of its contents using a target span (which could be stack or heap based). This involves a copy. I've not stackalloc-ed within the method as it would be unreadable as soon as it was returned.

If there were a spare 16 bytes before the start of the source span, it could be done in-place without the copy, though my main goal was avoiding excessive garbage collection whilst frequently calling a method that only took byte[] with variable length arrays. In hindsight, another option could have been to allocate a single byte[] for the worst case once, and to change the Length value under the hood to fit as needed - but this has been a fun diversion.

Anyhow - this seems to be working well for my use case, though I fear there are a few gotchas involving things like memory alignment still lurking. 

```
public static class UnsafeByteArrayExtension
{
    public static readonly nint ByteArrayTypeHandleValue = typeof(byte[]).TypeHandle.Value;
    public static readonly int NintSize = Unsafe.SizeOf<nint>();
    public static readonly int ByteArrayMinSize = Marshal.ReadInt32(ByteArrayTypeHandleValue, 4);
    public static readonly int ByteArrayLengthOffset = NintSize;
    public static readonly int ByteArrayDataOffset = NintSize + sizeof(uint) + sizeof(uint); // TypeHandle, Length and Padding values - needs tests.
    public static int GetUnsafeByteArraySize(this ReadOnlySpan<byte> span)
    {
        var rawDataSize = PadToNativeBoundary(span.Length);
        var byteArraySize = Math.Max(ByteArrayMinSize, rawDataSize + ByteArrayDataOffset);
        return byteArraySize;

        static int PadToNativeBoundary(int size)
        {
            var roundMask = NintSize - 1;
            return size + roundMask - ((size - 1) & roundMask);
        }
    }
    public static unsafe byte[] GetUnsafeByteArray(this ReadOnlySpan<byte> source, Span<byte> target)
    {
        Debug.Assert(source.GetUnsafeByteArraySize() <= target.Length);
        target.Cast<byte, nint>()[0] = ByteArrayTypeHandleValue;
        target[ByteArrayLengthOffset..].Cast<byte, uint>()[0] = (uint)source.Length;
        var memStart = (nint)Unsafe.AsPointer(ref target[0]);
        var unsafeArray = Unsafe.AsRef<byte[]>(&memStart);
        source.CopyTo(unsafeArray);
        return unsafeArray;
    }
}
```
